### PR TITLE
libcephfs: provide API to change umask

### PR DESCRIPTION
**Pull request by rpavani1998**
a binding for setting the umask.
[http://tracker.ceph.com/issues/24430](http://tracker.ceph.com/issues/24430)
Signed-off-by: Pavani Rajula <rpavani1998@gmail.com>

### CODE REVIEW

File: src/libcephfs.cc [hunk 1]
```diff
@@ -34,11 +34,15 @@
...
+#define DEFAULT_UMASK 002
+
+static mode_t umask_cb(void *);
 
 struct ceph_mount_info
 {
+  mode_t umask = DEFAULT_UMASK;
 public:
-  explicit ceph_mount_info(CephContext *cct_)
+   explicit ceph_mount_info(CephContext *cct_)
```
Comment (joscollin) [inline src/libcephfs.cc]: why change?

File: src/libcephfs.cc [hunk 2]
```diff
@@ -74,14 +78,19 @@ struct ceph_mount_info
   int init()
   {
     common_init_finish(cct);
-
     int ret;
+    struct client_callback_args args = [this]{
```
[the remark this answers is not in the document]
Comment (joscollin) [reply src/libcephfs.cc]: @tchaikov same thoughts here :)

[the remark this answers is not in the document]
Comment (batrick) [reply src/libcephfs.cc]: > please define args where it is used. also, why are you using this convoluted way to initialize args?

I suggested this code as a hack to get designated initializers for the struct with zero-initialization for the other members. I don't recall where I read that trick and now that I look at it again, I don't think it even does that. Got a better suggestion @tchaikov?

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/libcephfs.cc]: what about

```c++
client_callback_args args= {};
```

File: src/libcephfs.cc (hunk 1 above)
Comment (rpavani1998) [reply src/libcephfs.cc]: That was by mistake. Will revert that change.

File: src/libcephfs.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (batrick) [reply src/libcephfs.cc]: ya works for me. @rpavani1998 please do instead:

````
struct client_callback_args args = {};
args.handle = this
...
````

File: src/libcephfs.cc [hunk 3]
```diff
@@ -78,13 +78,11 @@ struct ceph_mount_info
...
     common_init_finish(cct);
+	  
     int ret;
-    struct client_callback_args args = [this]{
-        struct client_callback_args args;
-        args.handle = this;
-        args.umask_cb = umask_cb;
-        return args;
-    }();
+    struct client_callback_args args = {};
```
Comment (tchaikov) [inline src/libcephfs.cc]: again, please define `args` where is used. also, might want to drop `struct`, it's required by C, not C++.

Comment (rpavani1998) [reply src/libcephfs.cc]: If I declare `client_callback_args = {}` on line 116 It is giving me the error
`transfer of control bypasses initialization of: -- variable "args"`. so do I just have to initialize `args` on line 116?

Comment (batrick) [reply src/libcephfs.cc]: I think where it's defined is okay for the reason you stated. You can drop the "struct" though.

File: src/libcephfs.cc [hunk 4]
```diff
@@ -609,14 +634,14 @@ extern "C" int ceph_mkdir(struct ceph_mount_info *cmount, const char *path, mode
 {
   if (!cmount->is_mounted())
     return -ENOTCONN;
-  return cmount->get_client()->mkdir(path, mode, cmount->default_perms);
+  return cmount->get_client()->mkdir(path, cmount->umask ^ mode, cmount->default_perms);
```
Comment (batrick) [inline src/libcephfs.cc]: These umask operations should not be necessary. Please remove them.

File: src/libcephfs.cc [hunk 5]
```diff
@@ -609,14 +634,14 @@ extern "C" int ceph_mkdir(struct ceph_mount_info *cmount, const char *path, mode
...
 }
 
 extern "C" int ceph_mkdirs(struct ceph_mount_info *cmount, const char *path, mode_t mode)
 {
   if (!cmount->is_mounted())
     return -ENOTCONN;
-  return cmount->get_client()->mkdirs(path, mode, cmount->default_perms);
+  return cmount->get_client()->mkdirs(path,  cmount->umask ^ mode , cmount->default_perms);
```
Comment (batrick) [inline src/libcephfs.cc]: ditto

File: src/libcephfs.cc [hunk 6]
```diff
@@ -776,13 +801,13 @@ extern "C" int ceph_chmod(struct ceph_mount_info *cmount, const char *path, mode
 {
   if (!cmount->is_mounted())
     return -ENOTCONN;
-  return cmount->get_client()->chmod(path, mode, cmount->default_perms);
+  return cmount->get_client()->chmod(path, cmount->umask ^ mode, cmount->default_perms);
```
Comment (batrick) [inline src/libcephfs.cc]: ditto

File: src/libcephfs.cc [hunk 7]
```diff
@@ -776,13 +801,13 @@ extern "C" int ceph_chmod(struct ceph_mount_info *cmount, const char *path, mode
...
 }
 extern "C" int ceph_fchmod(struct ceph_mount_info *cmount, int fd, mode_t mode)
 {
   if (!cmount->is_mounted())
     return -ENOTCONN;
-  return cmount->get_client()->fchmod(fd, mode, cmount->default_perms);
+  return cmount->get_client()->fchmod(fd, cmount->umask ^ mode, cmount->default_perms);
```
Comment (batrick) [inline src/libcephfs.cc]: ditto

File: src/libcephfs.cc [hunk 8]
```diff
@@ -845,7 +870,7 @@ extern "C" int ceph_open(struct ceph_mount_info *cmount, const char *path,
 {
   if (!cmount->is_mounted())
     return -ENOTCONN;
-  return cmount->get_client()->open(path, flags, cmount->default_perms, mode);
+  return cmount->get_client()->open(path, flags, cmount->default_perms, cmount->umask ^ mode);
```
Comment (batrick) [inline src/libcephfs.cc]: ditto

File: src/libcephfs.cc [hunk 9]
```diff
@@ -107,6 +114,7 @@ struct ceph_mount_info
     if (ret)
       goto fail;
 
+    client->ll_register_callbacks(&args);
```
Comment (batrick) [inline src/libcephfs.cc]: @rpavani1998 another approach would be to do:

````
{
  client_callback_args args = {};
  args.handle = this;
  args.umask_cb = umask_cb;
  client->ll_register_callbacks(&args);
}
````

File: src/libcephfs.cc [hunk 10]
```diff
@@ -81,7 +85,7 @@ struct ceph_mount_info
       MonClient mc_bootstrap(cct);
       ret = mc_bootstrap.get_monmap_and_config();
       if (ret < 0)
-	return ret;
+	      return ret;
```
Comment (batrick) [inline src/libcephfs.cc]: @rpavani1998 let's remove this whitespace change. You can do:

```
# change your working tree to undo the changes in the last commit:
git diff HEAD HEAD^ | git apply
# now pick the changes you want to undo
git add -p
# Use `git reset` if you messed up and picked a change you don't want, then do `git add -p` again
# finally, amend your commit:
git commit --amend
````

File: src/libcephfs.cc [hunk 11]
```diff
@@ -616,7 +643,7 @@ extern "C" int ceph_mkdirs(struct ceph_mount_info *cmount, const char *path, mod
 {
   if (!cmount->is_mounted())
     return -ENOTCONN;
-  return cmount->get_client()->mkdirs(path, mode, cmount->default_perms);
+  return cmount->get_client()->mkdirs(path,  mode, cmount->default_perms);
```
Comment (batrick) [inline src/libcephfs.cc]: undo this whitespace change too please

**Pull request merged**